core/types/rlp_backwards_compat.libevm_test.go

```diff
@@ -179,14 +192,46 @@ func TestBodyRLPBackwardsCompatibility(t *testing.T) {
 					cmpeth.CompareHeadersByHash(),
 					cmpeth.CompareTransactionsByBinary(t),
 				}
-				if diff := cmp.Diff(body, got, opts); diff != "" {
-					t.Errorf("rlp.DecodeBytes(rlp.EncodeToBytes(%#v)) diff (-want +got):\n%s", body, diff)
+				if diff := cmp.Diff(want, gotWithoutExtra, opts); diff != "" {
+					t.Errorf("rlp.DecodeBytes(rlp.EncodeToBytes(%s)) diff (-want +got):\n%s", pretty.Sprint(want), diff)
 				}
 			})
 		})
 	}
 }
 
+// makeBody creates a [*Body] with the given arguments and ensures all exported
+// fields are set through the test [TestmakeBody].
+func makeBody(txs []*Transaction, uncles []*Header, withdrawals []*Withdrawal) *Body {
+	return &Body{
+		Transactions: txs,
+		Uncles:       uncles,
+		Withdrawals:  withdrawals,
+	}
+}
+
+func Test_makeBody(t *testing.T) {
```

Comment:
This is unnecessary due to the change-detector in the production file, which states that if it breaks then new backwards-compatibility tests are needed. That will naturally lead here.

(fyi) If this were to be kept (I don't think it should be), it could be simplified to only count the number of exported fields in body:

```go
var numExportedFields int
for i, typ := 0, reflect.TypeOf(&Body{}); i < typ.NumField(); i++ {
  if typ.Field(i).IsExported() {
    numExportedFields++
  }
}
assert.Equal(t, 3, numExportedFields, "makeBody() assumes this number of exported fields")
```

Reply:
I created this function because where it's called, the caller cannot use `body := Body{tx, u, w, nil}` because of the package name `vm_test` and `extras` being unexported. I feel like leaving it to `body := Body{Transactions: tx, ...}` is risky, what do you think?

Reply:
In the event that `geth` adds a new field, the change detector in `block.libevm.go` will no longer compile, and its comment says that new compatibility tests MUST be added, which exactly where you're using `makeBody()`. I think the indirection (forcing the reader to check what `makeBody()` does, thus breaking their flow) reduces the clarity of the test while the new function and its test don't add safety that we're not already achieving.

```go
// The RLP-related methods of [NOOPBodyHooks] make assumptions about the struct
// fields and their order, which we lock in here as a change detector. If this
// breaks then it MUST be updated and the RLP methods reviewed + new
// backwards-compatibility tests added.
```

Reply:
> If you feel strongly that this isn't enough

I do because that's the typical thing you would overlook and the test would still pass, at least for my past/future self. Good idea to move it to the types package, however it doesn't work due to `cmpeth` depending on the `types` package. I simplified the makeBody test in the meantime, feel free to suggest another solution though!

Reply:
I've refactored in #131 to avoid the need for this.

Reply:
I've refactored in #131 to avoid the need for this.